### PR TITLE
Fix weight mask interaction in objectives

### PR DESCRIPTION
**Pull request by wxs**
We used nonzero() on the weights in order to ensure that if there
happened to be a NaN or an Inf in the output that was going to be masked
about by the weights anyway, it wouldn't propagate (because 0*inf = NaN)
however this was causing interaction issues if you also used a mask,
because that wasn't using nonzero() properly.

This fixes that, and also fixes what I believe was an issue where I was
calling mean() instead of dividing by the sum of the sample weights.

Please think about this carefully: getting the normalization across both forms of weights right is a bit subtle, and I screwed up a number of times before creating this PR. I *think* it's right now though.

### CODE REVIEW

**Comment by wxs** [comment]
This fixes #573

File: keras/models.py [hunk 1]
```diff
@@ -64,19 +64,24 @@ def slice_X(X, start=None, stop=None):
...
-        # it's important that 0 * Inf == 0, not NaN, so I need to mask first
-        masked_y_true = y_true[weights.nonzero()[:-1]]
-        masked_y_pred = y_pred[weights.nonzero()[:-1]]
-        masked_weights = weights[weights.nonzero()]
-        obj_output = fn(masked_y_true, masked_y_pred)
+        '''
+        y_true dimension:  (sample, timestep, dims)
+        y_pred dimension:  (sample, timestep, dims)
+        weights dimension: (sample, timestep, 1)
+        '''
```
Comment (fchollet) [inline keras/models.py]: When a mask is provided, it means that the input has time steps. However the input might not necessarily be `(samples, timesteps, dims)`, it should more generally be considered to be `(samples, timesteps, **)` (e.g. sequences of pictures).

**Comment by fchollet** [comment]
Importantly, when you add tests, make them reproducible by seeding the RNG. I got:

```
E       AssertionError: assert 285.0 == 282.375
```

**Comment by fchollet** [comment]
Merged, thanks for the fix! I'll look at 3D support for sample weighting as well.

**Comment by wxs** [comment]
> Importantly, when you add tests, make them reproducible by seeding the RNG. I got:

@fchollet I saw that, though that's in a test that was already there!

**Pull request merged**